app-typescript/components/Message.tsx

```diff
@@ -0,0 +1,91 @@
+import * as React from 'react';
+import Text from './Text';
+import classNames from 'classnames';
+
+export type MessageProp =
+    | React.ReactNode
+    | string;
+
+export type Position = 'top' | 'bottom' | 'right' | 'left';
+
+export type  NotesType = 'default' | 'primary' | 'success' | 'warning' | 'alert' | 'highlight' | 'light';
+
+export interface IMessageOptions {
+    id: string;
+    avatar?: string;
+    duration?: number | null;
+    type?: NotesType;
+    position: Position;
+    icon?: string;
+}
+
+interface IProps extends IMessageOptions {
+    message: MessageProp;
+    position: Position;
+    closeElement(id, position): void;
+}
+
+export const Message = ({
+    id,
+    message,
+    type,
+    icon,
+    avatar,
+    duration,
+    position,
+    closeElement,
+}: IProps) => {
+    const container = React.useRef<HTMLDivElement | null>(null);
+    const [durationTime, setDurationTime] = React.useState(duration);
+    const intervalRef = React.useRef();
+
+    React.useEffect(() => {
```

Comment:
I think it's more readable code with using hooks, also it's a new feature of react and in the future, we will be able to maintain it easter, instead of rewriting setState in hooks, we are saving time with using it right now.

Reply:
Well, I don't know the hooks API so it's definitely not more readable to me. Likely to the rest of the team too since we aren't using it.

> we will be able to maintain it easter

What exactly do you think makes it easier to maintain compared to `setState` API?

> instead of rewriting setState in hooks, we are saving time with using it right now.

You mean one hook you just added? :) It would probably take an hour for you to change it, so it's not that significant.

Reply:
It's not just one if you look closely :) If I changed to setState instead of using hook than I should remove other hooks also. That implicates that I need to change the whole component logic with setting timers for appearing message and closing, also stopping the timer when the mouse enters a message, etc.  So it's more than an hour for me to change.

Reply:
Take a day to do it you need it or I can do it for you if you prefer so. I understand that it's a fun new API to play with, but it's a significant change in state management and we shouldn't making it without having concrete benefits in mind that we are doing this for.

Reply:
I'm up to using hooks, seems like it simplifies the code, not sure it's the case here but in general. it's more obvious if the component works with some api and ui state while here it was just single thing to take care of

Reply:
> seems like it simplifies the code

@petrjasek can you provide some examples where that would be the case? Personally, I have no difficulties working with `setState` API, so for me, it doesn't seem worth starting using it. We would burn a significant amount of developers' time until everyone picks it up and would end up with 2 approaches for the same problem without gaining anything except bragging rights about using the latest API :smile:

Reply:
In the future they will focusing on hooks instead of `setState` (newsletter received from them, something like simplifing code) so if we want to stay up-to-date with latest features I think we should start using new things. My vote is for hooks! Maybe better to start using them in UI Framework.

Reply:
it's all in [docs](https://reactjs.org/docs/hooks-intro.html), imo it's an investment to learn it not just burning devs time, it produces shorter and a bit more readable code keeping the related logic together (that's subjective though)

Reply:
What newsletter? I thought the react team only has a blog. What would that "focusing" even mean and how would it be relevant for us? Class components aren't deprecated and [there aren't any plans](https://reactjs.org/docs/hooks-intro.html#gradual-adoption-strategy) for this to change.

We should evaluate technology, be it APIs or libraries based on technical merit, not a release date.

There might be advantages to using hooks, as well as disadvantages. There's also a significant learning curve that would come with it. In our current situation, I'd rather spend the time refactoring existing code rather than introducing a new API which isn't necessary and it seems like we don't even know why we want it besides "it's new and better" argument.

Reply:
@petrjasek there's a lot in the docs which might all be relevant for different projects. For example, I don't buy "classes are confusing because you need to bind `this`" argument. Besides the binding thing, I don't see much confusion. Do you?

It also mentions optimizations that might be possible due to code folding, but I'd say it's the least of our issues. We have underperforming code, but it's mostly on the angular side. On react side, I'm not convinced that code folding(which isn't even available yet) would give us significant improvements over `shouldComponentUpdate`.

Regarding the shortness of the code, I wouldn't say it's a significant problem with class components. You need to bind methods, which does add boilerplate, but I'd say it's a minor inconvenience not worth worrying about.

Readable code is subjective as you say and given that we've been writing components without hooks for a few years, the thing that we are used to will probably be more readable. At least that's the case for me.

Keeping related logic together is important, but I'd say we are doing pretty good even without hooks, no?. I don't see it as that unrelated if on `componentDidMount` you start fetching data and are adding event listeners. Both are a part of the same component.

Reply:
I don't care much about this binding or possible optimizations, but simpler and better readable code is important imo. so in case of the toast, there is this bit for animation: 
```ts
class Component {
    constructor(props) {
        super(props);
        this.state = {isVisible: null};
    }

    componentDidMount() {
        this.setState({isVisible: true});
    }
}
```
with hooks it is
```ts
function Component() {
    const [isVisible, setIsVisible]  = useState(null);

    useEffect(() => setIsVisible(true), []);
}
```
and could be made reusable via:
```ts
const useIsVisible = () => {
     const [isVisible, setIsVisible]  = useState(null);

    useEffect(() => setIsVisible(true), []);

    return isVisible;
}

function Component {
    const isVisible = useIsVisible();
}
```
which I don't think we can do easily with class approach

Reply:
We can have reusable logic with classes by using higher-order components. One of the places it's used is in `connectCrudManager` and it has worked quite well so far. Here's an example:

```
function withVisibility(Component) {
    return class ComponentWithVisibility extends React.PureComponent {
        constructor(props) {
            super(props);
            this.state = {isVisible: null};
        }

        componentDidMount() {
            this.setState({isVisible: true});
        }

        render() {
            return (
                <Component visible={this.state.isVisible} />
            );
        }
    };
}
```

Initializing properties in the constructor is the same boring, but simple boilerplate as binding the properties, so I don't see how getting rid of it would add significant value.

I agree that reusability experience looks slightly nicer with hooks, but I wouldn't trade it for having to maintain 2 different approaches and the slowdown in development until developers pick it up and learn the gotchas.

If we limit it to `ui-framework` the downsides are less prominent and we may give it a try, but there's another issue that `refs` don't work with functional components. This may not be an issue with the toast but will be with others. We may wrap those, but the trouble doesn't seem worth it to me.

Reply:
I think refs work with [hooks](https://reactjs.org/docs/hooks-reference.html#useref)
anyhow starting with ui-framework makes sense, and it would also make sense to use just one approach, so what about hooks everywhere and avoiding classes @superdesk/client ?

Reply:
Refs do work with hooks, but probably not when used from class components. We may get it working, but I don't see a reason to put our time in this when the only significant upside we agree on is a slightly nicer syntax for reusability compared to higher-order functions.

Given there many real issues slowing down development which we'll need to address like the number and quality of UI components, refactoring, performance, moving services and features to extensions API, finishing the decoupling of planning, etc. I can't imagine why would we want to slow ourselves down even further by introducing another way of doing the same thing which already works **very well**. It may not be perfect, there might be space for improvement, but nonetheless the current class API is working very well.

Reply:
yep it all works fine, but what you might be missing is that it works very well for you, but for others it might be just easier to work with hooks because it's simpler and easier to read, thus maintain and make things reusable. but that's all subjective, functionality wise it's all the same. lets hear the others, if they don't care I would say we stick with classes as we have them all over the place, if they really care about using hooks in the framework I'm fine with that too. but we should make one or the other a rule so we keep things consistent

Reply:
also not sure about the slowdown, this hooks api is ~4 functions, or am I missing something?

Reply:
I don't know how many functions there are, but it's very different conceptually. You could compare it to generators vs promises or `setState` vs redux. You could say redux is only a few functions too, but it's different conceptually and takes a while to understand. It's not like it's just another lodash function that works independently in isolation. There would be many "how do I do x with hooks" and "why z doesn't work with hooks" questions everyone would need to figure out.

Reply:
> There would be many "how do I do x with hooks" and "why z doesn't work with hooks" questions everyone would need to figure out.

There **could** be..

I think conceptually it does the same thing like classes, it just changes the way it's written. There is no new/extra functionality.

created this [gist for alert components](https://gist.github.com/petrjasek/0512a7ad54e2173d6af016810f721832) so we talk about real usecase and not redux/lodash and other unrelated libs.

personally I can live with both, but with hooks I just find it easier to read

Reply:
> There could be..

Let's not be naive, there would **definitely** be.

The alert component even if used in production is basically a hello world with 10 elements without any lifecycle methods, event listeners, composition, network requests, async stuff, etc. so I don't think it helps any more than any other hello world example. You also forgot to take care of preventing re-renders which `PureComponent` version does. Memo has to be used with functional components.

> I think conceptually it does the same thing like classes, it just changes the way it's written.

I agree, but what I'm saying is that the change in the way it's written has a significant learning curve which would slow us down until we learn all the dos and don'ts of using hooks. I don't dislike the API, but I don't see a slight reduction of boilerplate worth the change.

If someone is very eager to use it, I'd suggest limiting it to ui-framework and only exporting class components to the outside.

Reply:
> If someone is very eager to use it, I'd suggest limiting it to ui-framework and only exporting class components to the outside.

fair enough, I would resolve this conversation then, ok?

Reply:
sure